### ISSUE TITLE
high RAM usage makes crash my Kobo.

### ISSUE THREAD

**Issue by anisg** [koreader/koreader#5998]
* KOReader version: v2020.03.2 , Semi-automated Method with KFMON.

* Device: Kobo Forma

#### Issue

After reading multiples epub for a few days, I start noticing a high usage of the RAM (reaching 140mo, when it start it's around 50mo), making the ereader slower when I change a page or when I navigate in the directories (5 times slower I say)

If I read a heavy book like a manga in CBZ format, the RAM usage get high really fast (reach 140mo after 50 pages of the manga). And eventually it crashes and I have to manually restart the ereader.

To prevent these issues, I usually restart KOReader from time to time. Anyone experienced a similar issues? Maybe a memory leaks?

**Comment by Frenzie** [comment]
Could you please include the crash.log?

Memory use is just cache so you can flip back to a previously rendered page really quickly. 140 MB isn't high on a device with 512 MB RAM.

**Comment by poire-z** [comment]
CBZ and PDF indeed cache bitmaps of previous pages in RAM, up to some limit.
But that is not the case with EPUBs.

We indeed leak memory. I start at 20M, and it gets slower when I reach 120M.
It increases a bit when switching document (but also somtimes decreases, so I guess it's not really the engines leaking).
It increases a lot when dict/wikipedia lookup (dict often triggering MuPDF to render HTML dict results).
It's also getting slower when switching to FM, or when closing History (when opened from the reader): we have a few calls to `collectgarbage()` on these occasions, and this takes more time when RAM is at 120M than when it is at 20M.. I tried removing them, but then I quite quickly reach 250M (while being quite more responsive - but then you get some spurious random micro slow down when the auto GC happens at any time) - and eventually some crash. Keeping them keeps me longer under 200M, with indeed some slowness.

So, I'm inclined to think it's Lua/Luajit memory that gets fragmented and badly re-used - so, it keeps increasing when more is needed for a new dict lookup or new book opened and engine initialising.
What I'd really like, but found no way to have it ensured, is: allocate 100M for lua stuff, and keep it working in that area. Then, have the engine (C code) allocate their stuff outside that, so these 100M don't get fragmented/filled by larger block of memory needed by the C engines...

The other option would be to not load C libraries in the main process, and have them done in a buddy process with interprocess communication :)

TLDR: yes, keep restarting KOReader from time to time when you see it reaches some 100-150M mem usage (that's what I do...)

**Comment by anisg** [comment]
Hi, sorry Frienzie I don't have a crash log, I just wanted to share a recurring issue I was having the last few months using KOReader.

It's the only issue I saw on KOreader, other than that it's great and so much better than Nickel.

I'm wondering why it gets low at 140mb of RAM, while the capacity of the Kobo Forma is 512mb. 

poire-z: I worked a few months with Lua and for me it always seemed to have a good memory management. I didn't get really low level though!

**Comment by poire-z** [comment]
@Frenzie @NiLuJe : no opinion, ideas, counter-theories,... about my soliloquy above ? :)

**Comment by NiLuJe** [comment]
I don't know if a valgrind run could shed some light on this, because Lua
;p.

**Comment by Frenzie** [comment]
@poire-z I haven't consciously noticed the slowdown behavior you describe. Generally speaking I don't think Lua(JIT) memory is really supposed to get fragmented by the allocator but I don't have anything meaningful to say about that, and if I tried it'd be armchair philosophy. :-)

I doubt I run KOReader without restarting for more than a few weeks at a time, so perhaps it's just not long enough for any potential issues to really manifest.

Since you mention the file manager might it bepossible that there are tables that are never dereferenced or something, thus causing a leak/slowdown?

**Comment by poire-z** [comment]
> Since you mention the file manager might it be possible that there are tables that are never dereferenced or something, thus causing a leak/slowdown?

Everything is possible :) but go find them... (and why more in the File Manager? that could happen anywhere)

**Comment by Frenzie** [comment]
I mean because you said you noticed it in the file manager, nothing more. ^_^

**Comment by NiLuJe** [comment]
Yeah, as far as my own experience goes, it's similar to @Frenzie. Except I only do ePubs, and even a week without a restart is extremely rare. And I don't document swap a lot.

**Comment by Frenzie** [comment]
Sometimes I'll only do EPUBs, sometimes I switch back & forth between EPUB and mostly PDF. I open & close a fair number of EPUBs these days thanks to Wallabag. Basically 5-50 page articles. (I can actually download my newspaper in EPUB and less interestingly in this case PDF but without automation I don't bother and just read it on my computer or my phone. )

I default my EPUBs to all the bells & whistles, which I assume takes the most memory. (Web render, best kerning, footnotes.)

I'm not a heavy dictionary user, but I'll do the occasional dictionary/Wikipedia lookup.

Right now after running a few days my KOReader says it's using 46 MB, that's in a proper 636 page book EPUB. That includes having previously opened a couple of PDFs and probably about a dozen EPUBs.

**Comment by poire-z** [comment]
> Right now after running a few days my KOReader says it's using 46 MB

I just restarted mine, started at 20M. A simple dict lookup involving a HTML dict and so bringing in MuPDF (although it might already be brought in for the InfoMessage icon in the "Opening..." initial popup) and I'm already at 60M ! That's with the first dict result being from Robert2007 dict, which has some fancy HTML.
Disabling it, doing the same workflow, first result being from HTML XMLLittre, and I'm only at 25M.
Disabling XMLLittre and first result being non-HTML Académie française dict and I'm at 24M.

I'm going to see how I go for some time with these HTML dicts disabled.

----

Just switching between 2 EPUBs (with a gesture, to not bring in TouchMenu code) also makes RAM usage grow up a bit.
Yesterday, I thought may be CreDocument self.buffer (buffer given to cre, 1.5M on my Kobo) being free() and re-alloc() on each opening might be the cause (1.5M freed, some small Lua object being put in that 1.5M and that 1.5M area is lost and the next 1.5M alloc needing a contiguous segment will allocate it elsewhere and increase RAM usage) - so, I just cached and re-used that self.buffer so it's not freed... but no real change in that small memory increase between switches... :(

**Comment by anisg** [comment]
Frienzie you have a Kobo or a Kindle? -- maybe it's a Kobo related issue.

I try different combinations and writed the RAM usage:

open 1 PDF:
- start koreader - ram 20MB
- open a 2.2MB PDF - ram 34MB 
- read 3 pages of the PDF - ram 66MB (+10MB each pages)
- read a few pages - ram 90/95MB (+5MB each pages, until it stabilize at 90/95MB)
- read a few pages - ram still 90/95MB  

open 1 epub:
- start koreader - ram 20MB
- open a 7MB EPUB - ram 43MB 
- read 1 page - ram 50MB  (+7 MB)
- read a few pages - ram 50MB (stabilized)
 
open 1 pdf and 1 epub:
- open pdf and read a few pages -> starts 34MB and stabilize at 90/95MB
- quit and go to menu
- open 1.1MB  epub and read a few pages -> starts at 114MB and stabilize at 140MB

open 2 epub:
- open first epub and read a few pages -> starts at 34MB, stabilize at 50MB
- open e epub and read a few pages -> starts at 69MB stabilize at 80MB 

A thing I noticed:
- the two epub above have lot of pages (1200 / 1800 pages). If I open a epub with less pages it adds less RAM.

**Comment by Frenzie** [comment]
That's on Kobo. You want to see slow, use Nickel. :-P

**Comment by poire-z** [comment]
Still upset with our memory leakage... Restarted yesterday at 20M, opened a few new books to find my next one, read a few pages, did a few lookups, and 15 hours later I'm at 220M...

https://github.com/koreader/koreader/pull/7857#discussion_r653665750
> Nope, it's not engine related, it's (probably) related to our inheritance system and anchored stale pointers everywhere ^^.

@NiLuJe : care to explain that a bit more ? Any idea about what we could do?

**Comment by Frenzie** [comment]
Do we currently have a way to see how much is in use as cache? I'd argue that to a very large extent that's likely to be highly desirable cache which enables reopening previously rendered pages really quickly.

**Comment by NiLuJe** [comment]
> Do we currently have a way to see how much is in use as cache? I'd argue that to a very large extent that's likely to be highly desirable cache which enables reopening previously rendered pages really quickly.

The (Document)Cache is capped at 64MB no matter what (which I actually think is an issue: I'd ideally like it capped much higher on devices with RAM to spare ^^).

**Comment by NiLuJe** [comment]
> > Nope, it's not engine related, it's (probably) related to our inheritance system and anchored stale pointers everywhere ^^.
> 
> @NiLuJe : care to explain that a bit more ? Any idea about what we could do?

That we have an insane amount of cyclical references to Document & Widget instances, as well as a large amount of extra references to them as upvalues because of our extensive use of closures, which makes tracking what's actually alive, where and why kinda tricky, and properly means a lot of it never gets GC'ed because there's a tiny (or not so tiny) sliver of something holding a reference to the whole thing left *somewhere*.

Not really sure how to go about it, I couldn't really find good tooling to investigate this properly :/.

**Comment by poire-z** [comment]
Note: I rarely read PDFs, so all my stuff is EPUB/crengine + MuPDF (without bitmap cache) for HTML lookups and images/covers.

**Comment by Frenzie** [comment]
@NiLuJe I had no idea what you were talking about but I realized the default is… different than my settings. The max should be something like 1 GB to prevent potential 32-bit issues but other than that I see no reason for any kind of max imho. The proportion of 0.4 works perfectly fine for that.

https://github.com/koreader/koreader/blob/2222f25422487e726db6c793de85af356e5fd493/defaults.lua#L35-L36

**Comment by NiLuJe** [comment]
> @NiLuJe I had no idea what you were talking about but I realized the default is… different than my settings. The max should be something like 1 GB to prevent potential 32-bit issues but other than that I see no reason for any kind of max imho. The proportion of 0.4 works perfectly fine for that.
> 
> https://github.com/koreader/koreader/blob/2222f25422487e726db6c793de85af356e5fd493/defaults.lua#L35-L36

Yeah, agreed, I think the hard-cap might have been an attempt at not overwhelming 128MB devices, but I haven't checked if it came before or after the 0.4 dynamic cap. In any case, screw 128MB devices ;p.

+1

**Comment by Frenzie** [comment]
0.4 * 128 = 51, so yeah, /shrug, can't really affect 128 MB devices can it ;-)

**Comment by TranHHoang** [comment]
Could we do something about it? I experience the same issue as well, after reading 2 books (which have about 10 images) and do some dictionary lookups, my RAM usage is already 280MB...

**Comment by Frenzie** [comment]
@NiLuJe 
> Not really sure how to go about it, I couldn't really find good tooling to investigate this properly :/. And OOP makes my brain hurt >_<".

Have you tried the ZeroBrane stack window? Basically something along these lines to insert the debugger:

```diff
diff --git a/base b/base
--- a/base
+++ b/base
@@ -1 +1 @@
-Subproject commit c35cb38e67e3441a438adff766fcab4dca1e99e7
+Subproject commit c35cb38e67e3441a438adff766fcab4dca1e99e7-dirty
diff --git a/reader.lua b/reader.lua
index 54d1793b..110bccfe 100755
--- a/reader.lua
+++ b/reader.lua
@@ -255,6 +255,10 @@ local function retryLastFile()
     }
 end
 
+package.path = package.path .. ";/opt/zbstudio/lualibs/mobdebug/?.lua"
+require('mobdebug').start()
+
+
 -- Start app
 local exit_code
 if file then
```

@TranHHoang 
> Could we do something about it? I experience the same issue as well, after reading 2 books (which have about 10 images) and do some dictionary lookups, my RAM usage is already 280MB.

Those aren't exclusive options. I see it climbing steadily only until it reaches a ceiling, which is expected although perhaps a little odd since we're talking about the same documents in this case. Similarly, a RAM usage of 280 MB is not an issue. You can always unnecessarily handicap your device and set the max cache to a lower value. Perhaps that's a bit of a requirement if you try to load documents that are significantly too high res for an embedded device for example?

But I assume you didn't end up here just because you were RAM-gazing. ^_^ Which kinds of settings do you use to read books? Which features do you enable? What exactly is the problem you mean by "it"?

**Comment by NiLuJe** [comment]
@Frenzie: It was on the list of things to try, and then I.. didn't ;p.

**Comment by Fell** [comment]
I think I'm also affected by this issue 
Device: Thalia Tolino Vision 5 (Physically identical with Kobo Libra H20, but different firmware)
Version: v2023.08 (installed using F-Droid Classic)

I'm often reading RPG rulebooks in PDF format which contain lots of images. Often, when I encounter a page with a large image the app just crashes. This seems to happen at or around 140-150 MB of memory usage (as shown on the status bar). With some PDFs I can reproduce this every time. I'm not having any issues on pages that contain mostly text. Could it be image-related?

Interestingly, if I use "Go to page" to jump to the image directly, It loads fine, and I can even flip to other pages after that. Just then it went up to and stabilized around 170 MB.

**Comment by NiLuJe** [comment]
You'll be even more memory constrained that on the matching Kobo on a Tolino, because Android ;o).
